src/utils.hpp

```diff
@@ -125,22 +131,88 @@ template <class Fn> class scope_exit {
     Fn fn_;
 };
 
+template <typename T>
+concept has_to_chars = requires(T v) { std::to_chars(nullptr, nullptr, std::declval<T>()); };
+
+template <typename T>
+concept has_from_chars = requires(T v) { std::from_chars(nullptr, nullptr, std::declval<T>()); };
+
 template <typename StringType, typename T>
 StringType to_string(T value)
-    requires std::is_integral_v<T>
+    requires std::is_integral_v<T> &&
+             (!std::is_same_v<T, bool>) && std::is_same_v<typename StringType::value_type, char> &&
+             std::is_same_v<StringType,
+                 std::basic_string<typename StringType::value_type,
+                     typename StringType::traits_type, typename StringType::allocator_type>>
 {
     // Maximum number of characters required to represent a 64 bit integer as a string
     // 20 bytes for UINT64_MAX or INT64_MIN + null byte
-    static constexpr size_t uint64_max_chars = 21;
+    static constexpr size_t max_chars = 21;
 
-    std::array<char, uint64_max_chars> str{};
+    std::array<char, max_chars> str{};
     auto [ptr, ec] = std::to_chars(str.data(), str.data() + str.size(), value);
-    if (ec == std::errc()) {
+    [[unlikely]] if (ec != std::errc()) {
+        return {};
+    }
+    return {str.data(), ptr};
+}
+
+template <typename StringType, typename T>
+StringType to_string(T value)
+    requires std::is_floating_point_v<T> && std::is_same_v<typename StringType::value_type, char> &&
+             std::is_same_v<StringType,
+                 std::basic_string<typename StringType::value_type,
+                     typename StringType::traits_type, typename StringType::allocator_type>>
+{
+    if constexpr (has_to_chars<double>) {
+        // 17 significant digits + dot + minus + exponent + minus + 3 exponend digits + null byte
+        static constexpr size_t max_chars = 25;
+
+        std::array<char, max_chars> str{};
+        auto [ptr, ec] = std::to_chars(str.data(), str.data() + str.size(), value);
+        [[unlikely]] if (ec != std::errc()) {
+            // This is likely unreachable if the max_chars calculation is accurate
+            return {};
+        }
         return {str.data(), ptr};
+    } else {
+        using char_type = typename StringType::value_type;
+        using traits_type = typename StringType::traits_type;
+        using allocator_type = typename StringType::allocator_type;
+        std::basic_ostringstream<char_type, traits_type, allocator_type> ss;
```

Comment:
would be closer to the `to_chars` version if you used [`setprecision(std::numeric_limits<T>::digits10)`](https://en.cppreference.com/w/cpp/io/manip/setprecision)

The difference being that `to_chars` may be smaller if the representation is unambiguous with fewer digits.